src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs

```diff
@@ -0,0 +1,327 @@
+
+    private async Task<Stream> GetAsyncStreamWithAuthRetry(string uriPath)
+    {
+        var response = await this.GetAsyncWithAuthRetry(uriPath);
+        return await response.Content.ReadAsStreamAsync();
+    }
+
+
+    private async Task EnsureTokenAsync()
+    {
+        if (this.client.DefaultRequestHeaders.Authorization == null)
+        {
+            await this.UpdateAuthTokenAsync();
+        }
+    }
+
+    private async Task UpdateAuthTokenAsync()
+    {
+        using (var activity = ActivitySource.StartActivity())
+        {
+            activity?.SetTag(QueryStrings.Username, this.Configuration.Username);
+
+            Log.LogInformation("Attempting to log in user {username}", this.Configuration.Username);
+
+            this.SetBasicAuthenticationHeader();
+            var response = await this.client.GetAsync(Paths.Login);
+
+            LoginResult? data = null;
+
+            if (response.IsSuccessStatusCode)
+            {
+                var json = await response.Content.ReadAsStringAsync() ?? String.Empty;
+
+                data = JsonSerializer.Deserialize<LoginResult>(json, options);
+            }
+
+            if (data == null)
+            {
+                Log.LogError("Login failed for user {username}.  Response: {response}", this.Configuration.Username, response);
+                throw new WattTimeClientException("Login failed.", response);
+            }
+
+            this.SetBearerAuthenticationHeader(data.Token);
+        }
+    }
+
+    private void SetBasicAuthenticationHeader()
+    {
+        var authToken = Encoding.ASCII.GetBytes($"{this.Configuration.Username}:{this.Configuration.Password}");
+        this.client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(AuthenticationHeaderTypes.Basic, Convert.ToBase64String(authToken));
+    }
+
+    internal void SetBearerAuthenticationHeader(string token)
+    {
+        this.client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(AuthenticationHeaderTypes.Bearer, token);
+    }
+
+    private async Task<string> MakeRequestAsync(string path, Dictionary<string, string> parameters, Dictionary<string, string>? tags = null)
+    {
+        using (var activity = ActivitySource.StartActivity(nameof(WattTimeClient)))
+        {
+            var url = BuildUrlWithQueryString(path, parameters);
+
+            Log.LogInformation("Requesting data using url {url}", url);
+
+            if (tags != null)
+            {
+                foreach (var kvp in tags)
+                {
+                    activity?.AddTag(kvp.Key, kvp.Value);
+                }
+            }
+
+            var result = await this.GetAsyncStringWithAuthRetry(url);
+
+            Log.LogDebug("For query {url}, received data {result}", url, result);
+
+            return result;
+        }
+    }
+
+    private string BuildUrlWithQueryString(string url, IDictionary<string, string> queryStringParams)
+    {
+        if (Log.IsEnabled(LogLevel.Debug))
```

Comment:
I assume this is done to avoid the `.Select` to be called?

Reply:
The string.Join call is potentially expensive, so no need to pay for it if debug logging is turned off.